### PR TITLE
BGPSec Router Certificate object class

### PR DESCRIPTION
**Pull request by jasdips**
And, some honing.

### CODE REVIEW

**Comment by jasdips** [comment]
Ties, Tom, I have added you as reviewers also, beside Andy. Please review if time permits. :)

**Comment by ties** [comment]
Hi Jasdip,

Here is a partial review :).

I had to double check the bgpsec certificate profile rfc (8209) since the
objects are not out there in the wild, but a bgpsec certificate can contain
one or more (including large ranges!) of ASNs,

3.1.3.5 <https://www.rfc-editor.org/rfc/rfc8209#section-3.1.3.5>. AS
Resources

   Each BGPsec Router Certificate MUST include the AS Resources
   extension, as specified in Section 4.8.11 of [RFC6487]
<https://www.rfc-editor.org/rfc/rfc6487#section-4.8.11>.  The
   AS Resources extension MUST include one or more ASNs, and the
   "inherit" element MUST NOT be specified.


I like the search by issuer (and not allowing that for any other object
type).

For bgpsec certificates we could consider searching by spki (?) since if i
recall correctly that (next to AS) is the only object in the output of
relying party software.

-Ties

File: draft-jasdips-regext-rdap-rpki.md [hunk 1]
```diff
@@ -50,25 +50,31 @@ end, RPKI defines the following cryptographic profiles:
...
 * Autonomous System Provider Authorization (ASPA, [@!I-D.ietf-sidrops-aspa-profile]) where an autonomous system number
   (ASN, [@!RFC5396]) holder cryptographically asserts about the provider AS for that ASN.
+* BGPSec Router Certificate ([@!RFC8209]) where an ASN holder cryptographically asserts that a router holding the
+  corresponding private key is authorized to emit secure route advertisements on behalf of the AS specified in the
+  certificate.
 
 This document defines a new RDAP extension, "rpki1", for accessing the RPKI registration data in the Regional Internet
-Registries (RIRs), including at national and local levels, for aforementioned RPKI profiles through RDAP. The intent is
-that such RDAP data can complement the existing RPKI diagnostic tools when troubleshooting a route hijack or leak, by
+Registries (RIRs), including at national and local levels, for aforementioned RPKI profiles through RDAP. The motivation
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: Isn't the terminology for all RIRs, LIRs, NIRs supposed to be Internet Number Registry (INR)? If this is intended for NIRs and/or LIRs, then INR is the right term not RIR.

File: draft-jasdips-regext-rdap-rpki.md [hunk 2]
```diff
@@ -50,25 +50,31 @@ end, RPKI defines the following cryptographic profiles:
...
-* Which IRR route object is related with a ROA?
-* Which ROAs are associated with an IP network?
+* Was a ROA created in conjunction with an Internet Routing Registry (IRR, [@RFC2622]) route?
+* Which IRR route is related with a ROA?
+* Which IP network is associated with a ROA?
 
-Beside the troubleshooting context, the ability to conveniently look up and search registered ROAs and ASPAs through
-RDAP would inform users, irrespective of their RPKI expertise level.
+Furthermore, correlating registered RPKI data with registered IP networks and autonomous system numbers would also give
+access to the latter's contact information through RDAP entity objects, which should come handy when troubleshooting.
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: OLD:
> which should come handy when troubleshooting

NEW:
> which should aid troubleshooting

File: draft-jasdips-regext-rdap-rpki.md [hunk 3]
```diff
@@ -110,7 +116,7 @@ The Route Origin Authorization (ROA) object class can contain the following memb
   representing the not-valid-after date of the end-entity certificate for the ROA ([@!RFC6487, section 4])
 * autoRenewed -- a boolean indicating if the registered ROA is auto-renewed or not
 * events -- see [@!RFC9083, section 4.5]
-* links -- links ([@!RFC9083, section 4.2]) for "self", and "related" to IP network and IRR objects
+* links -- links ([@!RFC9083, section 4.2]) for "self", and "related" to IP network and IRR (when defined) objects
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md:122]: maybe it is time to define the IRR objects. I would think only route and route6 are needed.

File: draft-jasdips-regext-rdap-rpki.md [hunk 4]
```diff
@@ -214,8 +220,9 @@ Searches for ROA information by origin autonomous system number are specified us
 
 rpkiRoas?originAutnum=AAAA
 
-AAAA is an autonomous system number representing the "originAutnum" property of a ROA, as described in (#roa_object_class).
-The following URL would be used to find information for ROAs with origin autonomous system number 65536:
+AAAA is an autonomous system number representing the "originAutnum" property of a ROA, as described in
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: AAAA is also the IPv6 DNS address record. Maybe just "AA" or use "XXX" as RFC 9082 does. Just a suggestion... not a big deal.

File: draft-jasdips-regext-rdap-rpki.md [hunk 5]
```diff
@@ -686,6 +696,230 @@ Here is an elided example for an autonomous system number object with ASPAs:
...
+    {
+      "description": [ "A BGPSec Router Certificate object in RDAP" ]
+    }
+  ]
+}
+```
+
+## Lookup
+
+The resource type path segment for exact match lookup of a BGPSec Router Certificate object is "rpkiBgpsecRouterCert".
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: See the latest additions to the extensions draft and how RFC 9082 addresses this... the path should either be "rpki1_bgpsecRouterCert" or "rpki1/bgpsecRouterCert".

personally I think the camel casing gets annoying and snake case is easier to read so "rpki1/bgpsec_router_cert" is how I find it easier to read.

File: draft-jasdips-regext-rdap-rpki.md [hunk 6]
```diff
@@ -686,6 +696,230 @@ Here is an elided example for an autonomous system number object with ASPAs:
...
+https://example.net/rdap/rpkiBgpsecRouterCert/65536
+```
+
+## Search
+
+The resource type path segment for searching BGPSec Router Certificate objects is "rpkiBgpsecRouterCerts".
+
+The following search path segments are defined for BGPSec Router Certificate objects:
+
+Syntax: rpkiBgpsecRouterCerts?handle=<handle search pattern>
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: see comments above on paths and extensions

File: draft-jasdips-regext-rdap-rpki.md [hunk 7]
```diff
@@ -686,6 +696,230 @@ Here is an elided example for an autonomous system number object with ASPAs:
...
+
+Syntax: rpkiBgpsecRouterCerts?issuer=<certificate issuer search pattern>
+
+Syntax: rpkiBgpsecRouterCerts?subject=<certificate subject search pattern>
+
+Searches for BGPSec router certificate information by handle are specified using this form:
+
+rpkiBgpsecRouterCerts?handle=XXXX
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: For search patterns, are you considering the partial string searching as specified in Section 4.1 of RFC 9082?

File: draft-jasdips-regext-rdap-rpki.md [hunk 8]
```diff
@@ -686,6 +696,230 @@ Here is an elided example for an autonomous system number object with ASPAs:
...
+  [
+    "rdap_level_0",
+    "rpki1",
+    "rpkiBgpsecRouterCert",
+    "rpkiBgpsecRouterCerts",
+    "rpkiBgpsecRouterCertSearchResults",
+    ...
+  ],
+  ...
+  "rpkiBgpsecRouterCertSearchResults":
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: see the latest in the extensions draft on search results and extensions.

File: draft-jasdips-regext-rdap-rpki.md [hunk 9]
```diff
@@ -789,6 +1031,36 @@ Contact: IETF <iesg@ietf.org>
 
 Intended usage: This extension identifier is used for accessing the RPKI registration data through RDAP.
 
+Extension identifier: rpkiBgpsecRouterCert
```
Comment (anewton1998) [inline draft-jasdips-regext-rdap-rpki.md]: Why are you doing multiple extension identifiers? Why not just "rpki1"?

**Review by anewton1998 — COMMENTED** [review]
Overall, I don't think you need to hammer out all these issues as long as you have some sort of note about them. For an -00 I think it is more important to get the basic ideas across.

**Comment by jasdips** [comment]
>I had to double check the bgpsec certificate profile rfc (8209) since the objects are not out there in the wild, but a bgpsec certificate can contain one or more (including large ranges!) of ASNs,
>
>3.1.3.5 <https://www.rfc-editor.org/rfc/rfc8209#section-3.1.3.5>. AS Resources
>
>   Each BGPsec Router Certificate MUST include the AS Resources extension, as specified in Section 4.8.11 of [RFC6487]
<https://www.rfc-editor.org/rfc/rfc6487#section-4.8.11>.  The AS Resources extension MUST include one or more ASNs, and the "inherit" element MUST NOT be specified.

That's a good catch. :) I also noticed it earlier in RFC 8209 but chose to start with just one ASN in the cert to prevent fate-sharing (as recommended for ROAs). But, on further thought, if an ISP holds multiple ASes in a registry and wants to create a single BGPSec router cert for all of them, that should be ok. I'll replace the "autnum" member with the "autnums" array of AS numbers.

Though, one operational question: Since ARIN does not currently support BGPSec Router Certificate profile, would you happen to know which peer RIRs presently support this profile?

> I like the search by issuer (and not allowing that for any other object type).

Yah, it makes sense here.

> For bgpsec certificates we could consider searching by spki (?) since if i recall correctly that (next to AS) is the only object in the output of relying party software.

OK. Are you envisioning a partial string search on the "publicKey" member?

**Comment by jasdips** [comment]
> Overall, I don't think you need to hammer out all these issues as long as you have some sort of note about them. For an -00 I think it is more important to get the basic ideas across.

Agreed. If we feel comfortable with the quality of this work, we could introduce it in regext after the next IETF. At least that's what I'm envisioning. :)

File: draft-jasdips-regext-rdap-rpki.md (hunk 1 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: > Isn't the terminology for all RIRs, LIRs, NIRs supposed to be Internet Number Registry (INR)? If this is intended for NIRs and/or LIRs, then INR is the right term not RIR.

Yes, will change.

File: draft-jasdips-regext-rdap-rpki.md (hunk 2 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: > OLD:
> 
> > which should come handy when troubleshooting
> 
> NEW:
> 
> > which should aid troubleshooting

OK.

File: draft-jasdips-regext-rdap-rpki.md (hunk 3 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md:122]: > maybe it is time to define the IRR objects. I would think only route and route6 are needed.

Impressed with you catching this. :) Right, we could explore that after this draft has been introduced in regext. 

Although, folks may ask why have it in RDAP when the IRRd Whois service exists. I think I know the answer (same rationale as for RDAP replacing Whois) but would involve some convincing.

File: draft-jasdips-regext-rdap-rpki.md (hunk 4 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: > AAAA is also the IPv6 DNS address record. Maybe just "AA" or use "XXX" as RFC 9082 does. Just a suggestion... not a big deal.

Good point. Easy enough to use another literal.

File: draft-jasdips-regext-rdap-rpki.md (hunk 7 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: > For search patterns, are you considering the partial string searching as specified in Section 4.1 of RFC 9082?

We should. Good point.

File: draft-jasdips-regext-rdap-rpki.md (hunk 9 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: > Why are you doing multiple extension identifiers? Why not just "rpki1"?

Duh, I still have a hangover from the multiple extensions used for the RIR Search but the rationale there was different. :) You are right. Will change.

**Comment by ties** [comment]
> > I had to double check the bgpsec certificate profile rfc (8209) since the objects are not out there in the wild, but a bgpsec certificate can contain one or more (including large ranges!) of ASNs,
> > 3.1.3.5 https://www.rfc-editor.org/rfc/rfc8209#section-3.1.3.5. AS Resources
> > Each BGPsec Router Certificate MUST include the AS Resources extension, as specified in Section 4.8.11 of [RFC6487]
> > https://www.rfc-editor.org/rfc/rfc6487#section-4.8.11.  The AS Resources extension MUST include one or more ASNs, and the "inherit" element MUST NOT be specified.
> 
> That's a good catch. :) I also noticed it earlier in RFC 8209 but chose to start with just one ASN in the cert to prevent fate-sharing (as recommended for ROAs). But, on further thought, if an ISP holds multiple ASes in a registry and wants to create a single BGPSec router cert for all of them, that should be ok. I'll replace the "autnum" member with the "autnums" array of AS numbers.

I do not think any RIR currently implements this profile. But this extension probably should try to model all the certificates that can be expressed in RPKI. Replacing it with an array (which can not easily contain `0-4294967296` feels safe to me since big ranges of ASNs allocated to a certificate are rare in practice.

I think rpki-client has this issue as well
```json
		{
			"asn": 15562,
			"ski": "BE889B55D0B737397D75C49F485B858FA98AD11F",
			"pubkey": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE4FxJr0n2bux1uX1Evl+QWwZYvIadPjLuFX2mxqKuAGUhKnr7VLLDgrE++l9p5eH2kWTNVAN22FUU3db/RKpE2w==",
			"ta": "ripe",
			"expires": 1732812001
		}
```

> Though, one operational question: Since ARIN does not currently support BGPSec Router Certificate profile, would you happen to know which peer RIRs presently support this profile?
> 
> > I like the search by issuer (and not allowing that for any other object type).
> 
> Yah, it makes sense here.
> 
> > For bgpsec certificates we could consider searching by spki (?) since if i recall correctly that (next to AS) is the only object in the output of relying party software.
> 
> OK. Are you envisioning a partial string search on the "publicKey" member?

I do not think so. I would want to exact match on hex SKI (like in the JSON).

File: draft-jasdips-regext-rdap-rpki.md (hunk 5 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: > See the latest additions to the extensions draft and how RFC 9082 addresses this... the path should either be "rpki1_bgpsecRouterCert" or "rpki1/bgpsecRouterCert".
> 
> personally I think the camel casing gets annoying and snake case is easier to read so "rpki1/bgpsec_router_cert" is how I find it easier to read.

Good point. Will change here and for other object classes.

File: draft-jasdips-regext-rdap-rpki.md (hunk 8 above)
Comment (jasdips) [reply draft-jasdips-regext-rdap-rpki.md]: Will change.

**Comment by jasdips** [comment]
> I do not think so. I would want to exact match on hex SKI (like in the JSON).

Good point. Will do.

**Pull request merged**